### ISSUE TITLE
27 gb csv file to csv

### ISSUE THREAD

**Issue by shariharan2002** [sqlitebrowser/sqlitebrowser#2855]
Hello all,
I am having a 16 column,150 million row 27 gb csv dataset.Its very raw(unclean) and it has to be cleaned before performing any analytics on it.
To clean,pandas is not facilitating me(memory error)So i went to SQL server but there its a whole other shebang(transaction log outgrowth etc.)
i have finally come to sqllite.i have downloaded the sqllite,sqllite studio and db browser for sqllit.I need to clean the data(many update and delete statements) and then keep the data ready for integration with powerbi.after struggling for almost 50-single handed man hours on trying to clean this, i am on the verge of giving up.But if anyone here could tell me that sqllite can handle it or if there are any specific ways/tricks for sqllite to handle such big datasets.pls tell me

my ultimate aim is to:
1.clean this
2.integrate to powerbi

pls help 
thanks

**Comment by chrisjlocke** [comment]
First off, split the 27 GB CSV into manageable chunks.  You're not going to be able to point DB4S at a 27 GB file and tell it to import it.  It'll have a go, but could take hours.  And then what happens if it fails half way?  Don't eat a horse by starting at the hooves.

I'd import the data by dumping it into DB4S.  You can then add columns, tidy, delete rows as necessary.  I wouldn't clean it before importing - tools to do that are just nasty.  But that's my opinion.

SQLite can handle 27 GB file, but it's obviously badly organised.  DB4S itself will struggle with tables that large - one of the first things it does is say "how many rows are in this table?" and if you have 27 GB in one table, that could take it a while to work out.  DB4S just isn't geared for that.  SQLite is though, and the command-line sqlite3.exe should be fine ... if you're OK with SQL.  Once the data is manageable, you can use DB4S again.  I tend to use DB4S to build the database and naked SQL to get data out...

If you need a tool to split that CSV, yell - it can be done in vb.net, python, etc.  Again, the GUI in DB4S won't really help much - it'll import it if it can, but you can't say 'only import 1 GB' or 'skip 1 GB then import 1 GB', etc.
Actually, thinking aloud, I might be able to help here.  If you do have a 27 GB CSV, I can knock up a tool to skip X lines, import X number of lines, etc.  That could be a useful tool...
Any chance of sending me the top 100 lines of the CSV?  Just so I know number of columns, field types, etc

**Comment by chrisjlocke** [comment]
> integrate to powerbi

I've not used powerBi, but just checked and it can use SQLite as a datasource.  You need to know that before importing 27 GB... 😉

**Comment by chrisjlocke** [comment]
> transaction log outgrowth etc

You're going to get something similar with SQLite.  Depending on the number of columns, etc, you're going to need to create indexes (preferably before importing) and these will take up space.  So a 27 GB CSV could take up more than that in SQLite database.

**Comment by shariharan2002** [comment]
Hi Chris
Thank you very very very much for your prompt reply and your offer to help!
I feel real happy when some expert is there to help me out.
Can you share your email so that I can send those csv images

**Comment by chrisjlocke** [comment]
> if you could elaborate a little more on the "fail on missing values"

Aah, I might have got this wrong.  Looks like it looks at the field definition, and if the field is set to not be null but the data being imported is null, then that's what this option is for.

![image](https://user-images.githubusercontent.com/7872218/136647670-fc0fbea0-5e10-4d68-b666-95ead3b4f1b1.png)

**Comment by shariharan2002** [comment]
> For working with a 20+ GB database like you're taking about, personally I wouldn't be using SQLite. Even though DB4S is a good gui tool.
> 
> The amount of reading and writing from disk would be huge (fast SSDs?), as would the memory requirements for the computer it's running on. 😱
> 
> If it was me doing it, and **if** the csv file is formatted correctly (so it can import without errors), personally I'd be using PostgreSQL to import it. Then adding indexes to all tables + fields I'd be likely to use for the clean up.
> 
> With fast SSDs and a bunch of memory, that should allow clean up operations to run in just a few seconds each, rather than 7 minutes per operation (!).
> 
> It sounds like you're more familiar with SQL Server than PostgreSQL though, so maybe use that?
> 
> Chris's advice to break up the csv into a number of small files, and import them one at a time first, is really good. That should help work out if the csv file can really import ok. And if it has issues with the import, then it'd help locate where the error is easier to fix it.

Sir i am not familiar with any SQL software!!!
i was trying to use SQL server, which is quite advanced to be used by a  beginner who just know simple SQL statements work
i came to know of db browser via dr chuck course on python -where he integrates a database to python
i thought i might give it a try
i have another doubt coming up pls help
i am posting it

**Comment by shariharan2002** [comment]
![image](https://user-images.githubusercontent.com/75315505/136647929-9fd7000f-7e06-4e61-b702-085f23a15798.png)


these are a set of sql statements i ran in the night(10 hrs)
it gives an errors and yes i did make an error 
in front of "origin" i did not put an equal to sign "="(my eyes got strained at night so lost focus sorry!)
so now after getting up in the morning i am seeing this error 
now the thing is after facing so many issues with SQL server for just executing 1 update statement
i am happy that has come to line 13

what i am wondering is ...has sqllite db browser executed the first 12 lines against the full dataset?
if yes ...then i am really happy
can you confirm this?

thanks a lot for all the help
i am really grateful for all your help right from my heart

i wish to receive this valuable and expert help until i integrate this to power bi(which i am rooting it to be successful)

again thanks a lot

**Comment by shariharan2002** [comment]
ok yes 
its writing a journal file

will the journal file be auto deleted after execution query stops

i am right now on my last statement 
and the journal file is 11 gb
so wth the execution of last statement its supposed to auto-delete correct
or can i delete it after the execution is over
is it safe?
since many people have said wrt SQL server that deleting transaction log is highly unsafe and you risk losing data
that s' why i am asking if manual deletion is safe after full execution of my current SQL statement?

tks a lot for all the help

**Comment by shariharan2002** [comment]
![image](https://user-images.githubusercontent.com/75315505/136650128-0543ad16-dcd9-4046-a524-0ab4b8f622b3.png)


in my table schema 
i have got for all colums the type as empty....since i disable auto data type detection on importing csv
so does that mean for all columns the default datatype is varchar...?

**Comment by shariharan2002** [comment]
![image](https://user-images.githubusercontent.com/75315505/136650173-7991a1c9-5ea4-4b71-8d59-00194ec6160e.png)
updated image

**Comment by shariharan2002** [comment]
ok my 1st of sql query statements is completed around 15 of them
this is assuming that the first 10 lines got executed in the night (See post 1hr ago and pls confirm)

**Comment by shariharan2002** [comment]
now i have 5 more sql delete statements  to be executed

![image](https://user-images.githubusercontent.com/75315505/136650376-64245126-d662-4a60-8762-6b52577a9afb.png)


the 1st 3 queries have to delete rows which are not numeric and the last 2 queries have to delete rows which are non-in date format

the above queries have beenwritten wrt SQL server
can you pls tell the appropriate functions for using these in sqllite dbbrowser

also could anyone tell if checked the trim command(see ystdy night post)...that means the string gets trimmed from both left and right side of string correct....why because there was no such option in SQL server and so i had to write custom SQL queries using RTRIM() and LTRIM() ....


also the journal log has not been deleted its occupying 11 gb ...i have completed 15 delete SQL queries 5 more to go (As referred above)...when will it disappear,...can i delete it manually or maybe if i click save database the journal log will disappear automatically?

also is there any way to check the total no of current rows(like is it in settings etc) present in database other than select count(*) from table
to avoid execution time


thanks a lot for all the help

**Comment by shariharan2002** [comment]
can you advice on what to do on not responding case?
can i close using task manager?
will the updates be written into database
as of now journal log is there
and database is also there in folder

so should it openthe updated database if i close the window?

**Comment by shariharan2002** [comment]
> It'll come back eventually! ;)
> 
> SQLite doesn't have an `isnumeric` function, so you have to rely on regular expressions. Nasty. https://stackoverflow.com/questions/32528759/how-to-check-if-a-value-is-a-number-in-sqlite/32528920
> 
> Again, this is where Python or vb.net can come into play. I can knock up a data cleaner app quite quickly later on based on this sort of query.

ok so you are asking me to stay put and not close it crt?

**Comment by shariharan2002** [comment]
i have stay put it has come back from not responding
thank god!!!!!

**Comment by shariharan2002** [comment]
ok 
so now having 5 delete statements ahead 
and already having executed 15 to 16 delete statement
and with journal log 11 gb and my c remaining space 20 gb

can i write the changes?
![image](https://user-images.githubusercontent.com/75315505/136651478-5c8aa5bd-b84f-47fd-89d9-194b94b0d152.png)

what i am worried is if click this button what are the possible scenarios can i again run into not responding mode and lose all the delete work that i have done since night ...is there maybe a safe way to do this...without computer slowing down?

thanks a lot for all the help

**Comment by shariharan2002** [comment]
![image](https://user-images.githubusercontent.com/75315505/136651700-999ff62b-36a8-4031-a462-cc070c6bd323.png)


current count of rows 144(original unclean database)-127(current count) =17 million rows(deleted) 
i have not yet clicked the write changes button
so 127 million rows is the initial imported version(due to fail on missing values it deleted 17 million rows) or the updated deleted version(i.e due to a combination of fail on missing values and my delete statement it has effectively deleted 17 million rows

basically now if i click write database will no of rows decrease or remain same at 127?

thanks a lot for all the help

**Comment by chrisjlocke** [comment]
> ok can you confirm about trim operations

I've not used that option but I would assume so.  Again, the easiest way to confirm is to pick some fields.
`select * from dataset where fieldX like ' %'`
The % is a placeholder for any text, so you are saying, 'give me all the rows where fieldX (change as required) starts with a space, then has any number of characters of anything'

It should come back with zero rows.

**Comment by chrisjlocke** [comment]
> and you told something about commit?

Commit is 'write changes'.  The other button 'Revert changes' is 'abort and don't write changes', also known as 'rollback'.
Surprised it only took seconds.  Yay.

> is that all? database has been updated?

Yup, good to go.

> but 1st column is not getting trimmed

It's hard to know for certain, but as the field contains quotation marks it may not be doing what you think.
Trim only removes spaces, so if the field is `...hello..` then it'll convert it to `hello`.  (spaces didn't show up well here, so using dots)

If your field is `"   test   "` (with quotes) then it won't convert that to `"test"` (ie, remove the spaces between the quotes)

**Comment by shariharan2002** [comment]
![image](https://user-images.githubusercontent.com/75315505/136654736-78a42fee-dfef-4d26-b264-b43510836285.png)


i just recreated the csv file by entering values again
its working it has trimmed all spaces under single quotes

so i think it should have done the same with my large dataset

so i think i will not run the select * statement since it shall consume a lot of time


now the only thing left according to me is verifying is numeric or not and is date or not?
trying to find out some way



thanks a lot for all the help
thank you very much

**Comment by chrisjlocke** [comment]
You have to set the name, salutation or number field to not accept null values.

![image](https://user-images.githubusercontent.com/7872218/136654987-f5b14483-7e1d-4dce-909f-471125cdcdd1.png)

Set the field to 'not null' by clicking the 'NN' column.  This means that column can't contain null values.
Just a nod that null is completely empty.  It does not mean an empty string.  Unsure if that makes sense, but yell if it doesn't.
So a field with 'not null' can still contain a value of '' (ie, an empty string).  An empty string isn't nothing.

**Comment by shariharan2002** [comment]
> Just a nod too, that you can't set the NN column if that field already contains nulls... DB4S doesn't like that as it violates the rules. So you can only set it on columns that already have values in them.

ok but this option is not available during import
and i ran those delete statements above(15 is null length etc)
so i am good to go ryt? as of no fields should contain the null character

and now with this 19 gb database
is there any advantage to modifying the table structure to NN ?is it time consuming?
is it safe now to do that an is it required especially since i have run delete statements for all columns?

**Comment by chrisjlocke** [comment]
> is there any advantage to modifying the table structure to NN ?is it time consuming?

Now that you've imported, no.

> ok but this option is not available during import

No, but when you import, you have the option to importing into an existing table.  So you can import one row to get the table structure, amend as needed, then import the data into this new table.

> so i am good to go ryt?

Yes.  if it was me (and I had the disk space) I would create a backup of the database now (even if you used 7zip to compress it) so you have a point in time to go back to.  You've come a long way and worked hard to get here.  It can only take a wrong delete statement to erase a lot of work.

Well done.  You've achieved a lot.

**Comment by shariharan2002** [comment]
ok chris 
i will take a backup now

one doubt
any way to check if the issue date and departure date is in date format
the date format used in these columns are yyyy-mm-dd
actually max no of rows should have been deleted by != 10 character length condition
but just want to verify
if i get a turnaround for this then i shall execute the date and numeric statements after my backup

thanks a lot for all the help

**Comment by chrisjlocke** [comment]
> and surprisingly it has recognized issue date and departure date as in date format as well ticketing airline cd, flight number and segment number as numeric?
this is absolutely amazing!!!!

This is only automatic detection for the plot.  Not for the database itself.

SQLite doesn't really bother with 'datatypes' so much - it's quite relaxed.  If you specify a field as integer, you can still insert 'fred' into it.  

> but in database structure i am not able to see those datatypes?

SQLite only really supports text, integer (whole numbers), real (float numbers), and blobs (binary data).

**Comment by shariharan2002** [comment]
a small issue
i tried importing my db to powerbi, turned it out to be memory error(c: has only 5gb free space sor currently freeing it and also i think powerbi is not exactly a "big" data analytics tool(i think)

but even if i were to go to another tool i realized 2 columns which are not required(marketing airline and ticketing airline.....since their respective "code" columns are present-->"marketing_airline_cd" and "ticketing_airline_cd"
so having taken a backup
i decided to execute a an alter table drop column marketing_airline

the journal file is growing and its growing fast!!!!
any idea how to control this?
what should i do??(all my freed space is now trying to get filled up)
is it possible to come up with the estimate journal size for 127 million rows if a field from each row is getting deleted 
like even a range will be helpful.

is there any way to avoid journal log only for this operation

and just wanted to know, can journal file growth crash the system since space is limited?

thanks  lot for all the help

**Comment by shariharan2002** [comment]
![image](https://user-images.githubusercontent.com/75315505/136672369-e920331d-4004-4d34-97b4-70f83f510a1a.png)
a prediction would be helpful since its at 16gb currently

**Comment by chrisjlocke** [comment]
> is there any way to avoid journal log only for this operation

You can turn the journal off by using `pragma journal_mode = off`

https://www.sqlite.org/pragma.html#pragma_journal_mode

> and just wanted to know, can journal file growth crash the system since space is limited?

If space is depleted, DB4S will abort the operation due to out of disk space.  It'll try and roll-back, but that will fail too, so yes, this could corrupt the database.

> is it possible to come up with the estimate journal size for 127 million rows if a field from each row is getting deleted
like even a range will be helpful.

The usual rule of thumb is that (depending on the operation) the journal could be an exact size of the database - if you perform an operation to delete every row in every database, to roll this back, it has to store all that data somewhere.
The issue is compounded if you then repeat the process without committing the journal - so there is no way to calculate the space needed - the space could be infinite.  If you edit a million rows continually without committing, that data has to sit somewhere.

I don't know your data, but if (for example) its primarily date based, you can split the data up in years into separate databases then get DB4S to attach these databases together.  They still won't be one database (no 'select * from dataset') but you can combine data across multiple databases for some queries.  This saves you having one 127 GB database and instead (for example) having 5 smaller databases.  Just a thought.

**Comment by chrisjlocke** [comment]
> i decided to execute a an alter table drop column marketing_airline

What this does is create a new table, then copy all the data from the old table to the new table.  If that table contained 90% of the data of your 19 GB database, then yes, the journal would be pretty large.  It then drops that table, which obviously fills up the journal some more.
Theoretically, if you have a backup, guess DB4S could have said, "wow - you've got a lot of data here - this could not only take a while but chew up your disk for the next hour ... hmm ... it appears you've only got 10 GB free ... sure you want to continue?!" or even turned off the journal.  Don't think that works 'on the fly' though - you'd have to close and re-open the database... but still a cool option to have on laaaaaaaarge databases.  I don't think most of our user base works with 19 GB databases.... ;)

**Comment by chrisjlocke** [comment]
> any suggestion for big data analytics software 

I've not used any, but lightly touched on Microsoft PowerBI.  As it's Microsoft there is a lot of tutorials and videos out there for it, so for support and hand-holding that's probably the best to go for.  You don't want to pick a product that you start struggling with but no one has heard of it so can't offer any help.  At least with PowerBI it's popular enough and well supported.*

* Well, yeah.  Huh huh.  If Visual Studio is anything to go by.  Turd of the century, that is.

**Comment by shariharan2002** [comment]
> > any suggestion for big data analytics software
> 
> I've not used any, but lightly touched on Microsoft PowerBI. As it's Microsoft there is a lot of tutorials and videos out there for it, so for support and hand-holding that's probably the best to go for. You don't want to pick a product that you start struggling with but no one has heard of it so can't offer any help. At least with PowerBI it's popular enough and well supported.*
> 
> .* Well, yeah. Huh huh. If Visual Studio is anything to go by. Turd of the century, that is.

ok 
but...uh can you suggest something other than powerbi that can satisfy 1 requirement
1.connect to large SQLlite databases and perform "quick" analysis/visualization?

**Comment by justinclift** [comment]
I'd probably look at something like Redash or Newdash for that:

* https://github.com/getredash/redash
* https://newdash.io

**Comment by shariharan2002** [comment]
So can i put it in like this?
![image](https://user-images.githubusercontent.com/75315505/136708260-ad9a9af6-ceab-4c9e-ad71-aa601a3d2656.png)

Please confirm
Thanks a lot for all the help!
